hal/armv7a/imx6ull/interrupts.c

```diff
@@ -0,0 +1,179 @@
+	/* 128 reserved */ gicd_icfgr0 = 0x700, /* 32 registers */ /* 32 reserved */ gicd_ppisr = 0x740, gicd_spisr0, /* 15 registers */
+	/* 112 reserved */ gicd_sgir = 0x7c0, /* 3 reserved */ gicd_cpendsgir = 0x7c4, /* 4 registers */ gicd_spendsgir = 0x7c8, /* 4 registers */
+	/* 40 reserved */ gicd_pidr4 = 0x7f4, gicd_pidr5, gicd_pidr6, gicd_pidr7, gicd_pidr0, gicd_pidr1, gicd_pidr2, gicd_pidr3,
+	gicd_cidr0, gicd_cidr1, gicd_cidr2, gicd_cidr3,
+
+	/* GIC virtual CPU interface register map */
+
+	gicv_ctlr = 0x800, gicv_pmr, gicv_bpr, gicv_iar, gicv_eoir, gicv_rpr, gicv_hppir, gicv_abpr, gicv_aiar, gicv_aeoir, gicv_ahppir
+	/* 41 reserved */, gicv_apr0 = 0x834, /* 3 reserved */ gicv_nsapr0 = 0x838, /* 6 reserved */ gicv_ciidr = 0x83f,
+	/* 960 reserved */ gicv_dir = 0xc00 };
+
+
+struct {
+	volatile u32 *gic;
+	intr_handler_t handlers[SIZE_INTERRUPTS];
+} interrupts_common;
+
+
+void interrupts_dispatch(void)
+{
+	u32 n, iar;
+
+	iar = *(interrupts_common.gic + gicv_iar);
+
+	/* Check that INT_ID is not 1023 or 1022 (spurious interrupt) */
+	if ((iar & 0x03fe) >= 0x03fe)
+		return;
+
+	n = iar & 0x1ff;
+
+	if (n >= SIZE_INTERRUPTS)
+		return;
+
+	if (interrupts_common.handlers[n].f != NULL)
+		interrupts_common.handlers[n].f(n, interrupts_common.handlers[n].data);
+
+	/* Update End of Interrupt register with original value from iar */
+	*(interrupts_common.gic + gicv_eoir) = (*(interrupts_common.gic + gicv_eoir) & 0x03ff) | iar;
+
+	return;
+}
+
+
+static void interrupts_enableIRQ(unsigned int irqn)
+{
+	*(interrupts_common.gic + gicd_isenabler0 + (irqn >> 5)) = 1 << (irqn & 0x1f);
+}
+
+
+static void interrupts_disableIRQ(unsigned int irqn)
+{
+	*(interrupts_common.gic + gicd_icenabler0 + (irqn >> 5)) = 1 << (irqn & 0x1f);
+}
+
+
+static void interrupts_setConf(unsigned int irqn, u32 conf)
+{
+	u32 t;
+
+	t = *(interrupts_common.gic + gicd_icfgr0 + (irqn >> 4)) & ~(0x3 << ((irqn & 0xf) << 1));
+	*(interrupts_common.gic + gicd_icfgr0 + (irqn >> 4)) = t | ((conf & 0x3) << ((irqn & 0xf) << 1));
+}
+
+
+static void interrupts_setPriority(unsigned int irqn, u32 priority)
+{
+	u32 mask = *(interrupts_common.gic + gicd_ipriorityr0 + (irqn >> 2)) & ~(0xff << ((irqn & 0x3) << 3));
+
+	*(interrupts_common.gic + gicd_ipriorityr0 + (irqn >> 2)) = mask | ((priority & 0xff) << ((irqn & 0x3) << 3));
+}
+
+
+static inline u32 interrupts_getPriority(unsigned int irqn)
+{
+	return (*(interrupts_common.gic + gicd_ipriorityr0 + (irqn >> 2)) >> ((irqn & 0x3) << 3)) & 0xff;
+}
+
+
+int hal_interruptsSet(unsigned int n, int (*f)(unsigned int, void *), void *data)
+{
+	if (n >= SIZE_INTERRUPTS)
+		return -1;
+
+	hal_interruptsDisable();
+	interrupts_common.handlers[n].data = data;
+	interrupts_common.handlers[n].f = f;
+
+	if (f == NULL) {
+		interrupts_disableIRQ(n);
+	}
+	else {
+		interrupts_setPriority(n, 0xa);
+		interrupts_setConf(n, 0x3);
+		interrupts_enableIRQ(n);
+	}
+
+	hal_interruptsEnable();
+
+	return 0;
+}
+
+
+static void interrupts_gicSetAddr(void)
```

Comment:
I think that `initAddr` would be better name. `Set` suggest that this function is a setter (it is not).

Reply:
Yeah, it would be better. Fixed